src/plonk/prover.rs

```diff
@@ -20,256 +20,399 @@ use crate::transcript::TranscriptWrite;
+                        )
+                    })
+                    .collect()
+            },
+        )
+        .collect();
+
+    let lookups_vec = match lookups_vec {
+        Ok(lookups_vec) => lookups_vec,
+        Err(err) => return Err(err),
+    };
 
     // Sample beta challenge
     let beta = ChallengeBeta::get(transcript);
 
     // Sample gamma challenge
     let gamma = ChallengeGamma::get(transcript);
 
-    // Commit to permutations, if any.
-    let permutations = pk
-        .vk
-        .cs
-        .permutations
+    let permutations_vec: Result<Vec<Vec<_>>, _> = advice_vec
         .iter()
-        .zip(pk.permutations.iter())
-        .map(|(p, pkey)| p.commit(params, pk, pkey, &witness.advice, beta, gamma, transcript))
-        .collect::<Result<Vec<_>, _>>()?;
+        .map(
+            |advice| -> Result<Vec<permutation::prover::Committed<C>>, Error> {
+                // Commit to permutations, if any.
+                pk.vk
+                    .cs
+                    .permutations
+                    .iter()
+                    .zip(pk.permutations.iter())
+                    .map(|(p, pkey)| {
+                        p.commit(
+                            params,
+                            pk,
+                            pkey,
+                            &advice.advice_values,
+                            beta,
+                            gamma,
+                            transcript,
+                        )
+                    })
+                    .collect()
+            },
+        )
+        .collect();
 
-    // Construct and commit to products for each lookup
-    let lookups = lookups
+    let permutations_vec = match permutations_vec {
+        Ok(permutations_vec) => permutations_vec,
+        Err(err) => return Err(err),
+    };
+
+    let lookups_vec: Result<Vec<Vec<lookup::prover::Committed<'_, C>>>, _> = lookups_vec
         .into_iter()
-        .map(|lookup| lookup.commit_product(&pk, &params, theta, beta, gamma, transcript))
-        .collect::<Result<Vec<_>, _>>()?;
+        .map(|lookups| -> Result<Vec<_>, _> {
+            // Construct and commit to products for each lookup
+            lookups
+                .into_iter()
+                .map(|lookup| lookup.commit_product(&pk, &params, theta, beta, gamma, transcript))
+                .collect::<Result<Vec<_>, _>>()
+        })
+        .collect();
+
+    let lookups_vec = match lookups_vec {
+        Ok(lookups_vec) => lookups_vec,
+        Err(err) => return Err(err),
+    };
 
     // Obtain challenge for keeping all separate gates linearly independent
     let y = ChallengeY::get(transcript);
 
-    // Evaluate the h(X) polynomial's constraint system expressions for the permutation constraints, if any.
-    let (permutations, permutation_expressions): (Vec<_>, Vec<_>) = {
-        let tmp: Vec<_> = permutations
+    let (permutations_vec, permutation_expressions_vec): (Vec<Vec<_>>, Vec<Vec<_>>) =
```

Comment:
There might be a better way to do this relying upon iterators rather than concrete vectors.